Fixed ExternalParser inside of WaitingExternalParser

ahmedowian opened the pull request:
Please see if this pull request is fine for our fix.
@rgauss

ahmedowian committed:
Testing MM

ahmedowian committed:
MM-292: Refresh of Technical Metadata after Trim is Inconsistent
   - Used Thread.join to fix the problem.
   - TODO:
     Copy patched version of ExternalParser into WaitingExternalParser
     Cleanup code to remove printlns
     Perform full builds of all dependencies and integration test
     Submit ticket to tika along with fix
     Submit unit test to tika to demonstrate bug. Use cat instead of ffmpeg

ahmedowian committed:
MM-292: Refresh of Technical Metadata after Trim is Inconsistent
   - Copied patched version of ExternalParser into WaitingExternalParser
   - Cleaned up code to remove printlns
   - TODO:
     Perform full builds of all dependencies and integration test
     Submit ticket to tika along with fix
     Submit unit test to tika to demonstrate bug. Use cat instead of ffmpeg

ahmedowian committed:
MM-292: Refresh of Technical Metadata after Trim is Inconsistent
   - Cleaned up code and updated pom
   - TODO:
     Perform full builds of all dependencies and integration test
     Submit ticket to tika along with fix
     Submit unit test to tika to demonstrate bug. Use cat instead of ffmpeg

rgauss commented on src/main/java/org/apache/tika/parser/external/WaitingExternalParser.java:
```diff
@@ -15,48 +13,362 @@
+            }
+            catch (InterruptedException e)
+            {
+                e.printStackTrace();
```
This would be thrown when the process was asked to stop so we don't need to print stack trace.

We may want to call `parser.interrupt()` though.  I don't recall if the interrupt state propagates to spawned new threads like parser here.

ahmedowian replied:
Yeah, I actually meant to yank that out and put a do-nothing comment.

ahmedowian committed:
MM-292: Refresh of Technical Metadata after Trim is Inconsistent
   - Removed e.printStacktrace()
   - TODO:
     Perform full builds of all dependencies and integration test
     Submit ticket to tika along with fix
     Submit unit test to tika to demonstrate bug. Use cat instead of ffmpeg

ahmedowian committed:
MM-292: Refresh of Technical Metadata after Trim is Inconsistent
   - Handled parser threads after waiting.
   - TODO:
     Perform full builds of all dependencies and integration test
     Submit ticket to tika along with fix
     Submit unit test to tika to demonstrate bug. Use cat instead of ffmpeg

rgauss commented on src/main/java/org/apache/tika/parser/external/WaitingExternalParser.java:
```diff
@@ -15,48 +13,391 @@
+    /*
+     * A minute to wait for parsing seems more than enough time, but it's definitely better than forever.
+     */
+    private static final long MAX_WAIT_TIME_MS = 60 * 1000;
```
This should be more descriptive, something like `MAX_STREAM_PARSE_WAIT_TIME_MS`

rgauss commented on src/main/java/org/apache/tika/parser/external/WaitingExternalParser.java:
```diff
@@ -15,48 +13,391 @@
+           process = Runtime.getRuntime().exec( cmd );
+        }
+
+        Parser stdoutParser = null;
```
Calling these parsers (within a class called *Parser) is going to lead to confusion.  Perhaps `stdoutStreamParser` and `stderrStreamParser`.

rgauss commented on src/main/java/org/apache/tika/parser/external/WaitingExternalParser.java:
```diff
@@ -15,48 +13,391 @@
+            try {
+                process.waitFor();
+            } catch (InterruptedException ignore) {
+            }
```
Here we need to check whether or not `stdoutStreamParser` and `stderrStreamParser` are null, and if not call `interrupt()` on them.

rgauss commented on src/main/java/org/apache/tika/parser/external/WaitingExternalParser.java:
```diff
@@ -15,48 +13,391 @@
+        waitForParserToFinish(stderrParser);
+    }
+    
+    private void waitForParserToFinish(Parser parser) throws IOException
```
Similar to above, should probably be something like `waitForStreamParserToFinish` with a `streamParser` argument.

rgauss commented on src/main/java/org/apache/tika/parser/external/WaitingExternalParser.java:
```diff
@@ -15,48 +13,391 @@
+            }
+            catch (InterruptedException ignore)
+            {
+                // This happens only if client code calls interrupt on this thread.
```
In this case we don't want the stream parsing thread to continue.  `interrupt()` should be called on it.  See https://github.com/rgauss/thread-interrupt-test

rgauss commented on src/main/java/org/apache/tika/parser/external/WaitingExternalParser.java:
```diff
@@ -15,48 +13,391 @@
+               } finally {
+                  IOUtils.closeQuietly(reader);
+                  IOUtils.closeQuietly(stream);
+                  this.isDone = true;
```
We could have gotten here after an IOException so this isn't a true measure of 'done'.

rgauss commented on src/main/java/org/apache/tika/parser/external/WaitingExternalParser.java:
```diff
@@ -15,48 +13,391 @@
+            }
+            else
+            {
+                parser.interrupt();
```
If `join` kills the thread I don't think we need to interrupt it here.

rgauss commented on src/main/java/org/apache/tika/parser/external/WaitingExternalParser.java:
```diff
@@ -15,48 +13,391 @@
+            else
+            {
+                parser.interrupt();
+                throw new IOException("The parser took too long to parse the output from the external command.");
```
This should be a `TimeoutException` and perhaps mention the timeout value: "... took longer than XXXms".  The `parse` method can catch that and rethrow as a `TikaException` if need be.

rgauss commented:
I'll merge then make these changes.

rgauss commented:
Merged, made changes discussed above, plus a few other minor ones.
